robots/pkg/test-label-analyzer/stats.go

```diff
@@ -0,0 +1,138 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package test_label_analyzer
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// FileStats contains the information of the file whose outline was traversed and the results of the
+// traversal.
+type FileStats struct {
+	*Config    `json:"config"`
+	*TestStats `json:"test_stats"`
+
+	// RemoteURL is the absolute path to the file, most certainly an absolute URL inside a version control repository
+	// containing a commit ID in order to exactly define the state of the file that was traversed
+	RemoteURL string `json:"path"`
+}
+
+// TestStats contains the results of traversing a set of Ginkgo outlines and collecting the pathes in the form of
+// GinkgoNode slices matching a Config describing the criteria to match against.
+type TestStats struct {
+
+	// SpecsTotal is the total number of specs encountered during traversal of the outline
+	SpecsTotal int `json:"specs_total"`
+
+	// MatchingSpecPathes is the slice of PathStats to matching specs for the collection of outlines traversed.
+	// Each PathStats inside this slice is the path to each of the matching specs defined by the Config
+	// being used. Please note that the GinkgoNode.Nodes are being removed during traversal.
+	MatchingSpecPathes []*PathStats `json:"matching_spec_pathes"`
+}
+
+// PathStats contains all relevant data to a path matching a Config.
+type PathStats struct {
+
+	// Lines has the line numbers for the matching nodes
+	Lines []int `json:"lines"`
+
+	// GitBlameLines is the output of the blame command for each of the Lines
+	GitBlameLines []*GitBlameInfo `json:"git_blame_lines"`
+
+	// Path denotes the path to the spec that has been found to match
+	Path []*GinkgoNode `json:"path"`
+}
+
+func GetStatsFromGinkgoOutline(config *Config, gingkoOutline []*GinkgoNode) *TestStats {
+	stats := &TestStats{
+		SpecsTotal: 0,
+	}
+	traverseNodesRecursively(stats, config, gingkoOutline, nil)
+	return stats
+}
+
+func traverseNodesRecursively(stats *TestStats, config *Config, gingkoOutline []*GinkgoNode, parents []*GinkgoNode) {
+	for _, node := range gingkoOutline {
+		var parentsWithNode []*GinkgoNode
+		parentsWithNode = append(parentsWithNode, parents...)
+		parentsWithNode = append(parentsWithNode, node)
+		if node.Spec {
+			stats.SpecsTotal++
+			for _, category := range config.Categories {
+				var testName string
+				for _, nodeFromPath := range parentsWithNode {
+					testName = strings.Join([]string{testName, nodeFromPath.Text}, " ")
+				}
+				if category.TestNameLabelRE.MatchString(testName) {
+					var path []*GinkgoNode
+					for _, pathNode := range parentsWithNode {
+						path = append(path, pathNode.CloneWithoutNodes())
+					}
+					stats.MatchingSpecPathes = append(stats.MatchingSpecPathes, &PathStats{
+						Path: path,
+					})
+				}
```

Comment:
Thinking out loud: an improvement could be to pass a boolean to the depth call which tells if the parent already match. I mean if parent node path matches the regex => we can be sure that all the child nodes match as well.
Something like:
```
func traverseNodesRecursively(stats *TestStats, config *Config, gingkoOutline []*GinkgoNode, parents []*GinkgoNode, matched bool) {
	for _, node := range gingkoOutline {
		var parentsWithNode []*GinkgoNode
		parentsWithNode = append(parentsWithNode, parents...)
		parentsWithNode = append(parentsWithNode, node)
		if node.Spec {
			stats.SpecsTotal++
			for _, category := range config.Categories {
				if matched {
					addMatchingSpecPaths(stats, parentsWithNode)
				} else {
					var testName string
					for _, nodeFromPath := range parentsWithNode {
						testName = strings.Join([]string{testName, nodeFromPath.Text}, " ")
					}
					if category.TestNameLabelRE.MatchString(testName) {
						matched = true
						addMatchingSpecPaths(stats, parentsWithNode)
					}
				}
			}
		}
		traverseNodesRecursively(stats, config, node.Nodes, parentsWithNode, matched)
	}
}

func addMatchingSpecPaths(stats *TestStats, nodes []*GinkgoNode) {
	var path []*GinkgoNode
	for _, pathNode := range nodes {
		path = append(path, pathNode.CloneWithoutNodes())
	}
	stats.MatchingSpecPathes = append(stats.MatchingSpecPathes, &PathStats{
		Path: path,
	})
}
```
But I am not sure whether it is worth it.

Reply:
I think it's not worth it since we only want to collect the Spec nodes, all child nodes of Spec nodes I have seen do have `node.Spec == false`, so I guess it's not. Let's defer this optimization until we see we need it. WDYT?

Reply:
Sure :) As I said, I was thinking out loud